### PR TITLE
Extract validator into separate package

### LINKED ISSUE #466: Extract validator in separate package
Validator is quite useful on it's own so it should be separated in separate package.

Issues to be discussed:
- logger sharing with lodestar
- rpc api interfaces

### PR DESCRIPTION
**Pull request by mpetrunic**
- a lot of changes
- validator is now separate package

resolves #466

### CODE REVIEW

**Review by GregTheGreek — DISMISSED** [review]

File: packages/lodestar-validator/README.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# Lodestar Validator
+[![Discord](https://img.shields.io/discord/593655374469660673.svg?label=Discord&logo=discord)](https://discord.gg/aMxzVcr)
+![ETH2.0_Spec_Version 0.8.3](https://img.shields.io/badge/ETH2.0_Spec_Version-0.8.3-2e86c1.svg)
+
+Typescript imlementation of the Eth2.0 beacon chain
```
Comment (wemeetagain) [inline packages/lodestar-validator/README.md]: Readme should probably be stripped down for validator

File: packages/lodestar-validator/src/util/beaconChain.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
+import {IBeaconConfig} from "@chainsafe/eth2.0-config";
```
Comment (wemeetagain) [inline packages/lodestar-validator/src/util/beaconChain.ts:1]: makes me sad :sob:

File: packages/lodestar/src/validator/nodejs.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
+import defaultConfig, {IValidatorClientOptions} from "./options";
```
Comment (wemeetagain) [inline packages/lodestar/src/validator/nodejs.ts:1]: Probably for a different PR:
I think this should live in the validator package.
Looks like we will probably need to split out our lodestar modules even more if we want to do that though.

File: packages/lodestar/test/unit/util/objects.test.ts [hunk 4]
```diff
@@ -1,28 +1,12 @@
-// @ts-ignore-file
 import BN from "bn.js";
 import {expect} from "chai";
-
 import {uint64} from "@chainsafe/eth2.0-types";
 import {config} from "@chainsafe/eth2.0-config/lib/presets/mainnet";
-import {isPlainObject, mostFrequent} from "../../../src/util/objects";
-
+import {mostFrequent} from "../../../src/util/objects";
+import {describe, it} from "mocha";
```
Comment (wemeetagain) [inline packages/lodestar/test/unit/util/objects.test.ts:6]: is this going in all of our tests now?

**Review by wemeetagain — CHANGES_REQUESTED** [review]
roughly lgtm!
Makes me realize we need to split out the lodestar modules into packages to make things 'right'.

File: packages/lodestar-validator/README.md (hunk 1 above)
Comment (mpetrunic) [reply packages/lodestar-validator/README.md]: oh yeah, true

File: packages/lodestar-validator/src/util/beaconChain.ts (hunk 2 above)
Comment (mpetrunic) [reply packages/lodestar-validator/src/util/beaconChain.ts:1]: I know :(

File: packages/lodestar/src/validator/nodejs.ts (hunk 3 above)
Comment (mpetrunic) [reply packages/lodestar/src/validator/nodejs.ts:1]: I think about it a lot. Probably it's much nicer UX to do `yarn add lodestar` and immediately have standalone validator. I stripped down validator package to be unbiased on logger and database implementation.

File: packages/lodestar/test/unit/util/objects.test.ts (hunk 4 above)
Comment (mpetrunic) [reply packages/lodestar/test/unit/util/objects.test.ts:6]: Dunno, cli works without it but I got Visual Code and Webstrom complaining when it's missing

File: packages/lodestar-validator/src/util/beaconChain.ts (hunk 2 above)
Comment (ec2) [reply packages/lodestar-validator/src/util/beaconChain.ts:1]: :(

Comment (GregTheGreek) [reply packages/lodestar-validator/src/util/beaconChain.ts:1]: wait whats going on here...?

Comment (mpetrunic) [reply packages/lodestar-validator/src/util/beaconChain.ts:1]: haha. I had to copy few state transition methods in this file as we cannot easily rip state transition into separate module

File: packages/lodestar/package.json [hunk 5]
```diff
@@ -2,7 +2,7 @@
   "name": "@chainsafe/lodestar",
```
Comment (ec2) [inline packages/lodestar/package.json:2]: Thoughts on renaming this to `lodestar-chain` or `lodestar-beacon` or something like that?

Comment (mpetrunic) [reply packages/lodestar/package.json:2]: reason?

Comment (GregTheGreek) [reply packages/lodestar/package.json:2]: tbh - i kind of back this idea to some degree... Lodestar has sort of evolved from this massive beast, to a small army of individual components.

File: packages/lodestar/src/cli/commands/validator.ts [hunk 6]
```diff
@@ -2,13 +2,11 @@
  * @module cli/commands
```
Comment (ec2) [inline packages/lodestar/src/cli/commands/validator.ts:2]: Maybe for another PR, but we should totally have the CLI in a different package and have that require both the beacon chain and the validator.

File: packages/lodestar/package.json (hunk 5 above)
Comment (ec2) [reply packages/lodestar/package.json:2]: ^ That. Lodestar will eventually include the light client or something too. Helps distinguish what is what

File: packages/lodestar/src/cli/commands/validator.ts (hunk 6 above)
Comment (GregTheGreek) [reply packages/lodestar/src/cli/commands/validator.ts:2]: Yeah but here's a problem, debugging on the fly could be a pain. You'd have to keep lerna linking no?

File: packages/lodestar/test/unit/util/objects.test.ts (hunk 4 above)
Comment (GregTheGreek) [reply packages/lodestar/test/unit/util/objects.test.ts:6]: same for me

File: packages/lodestar/package.json (hunk 5 above)
Comment (wemeetagain) [reply packages/lodestar/package.json:2]: I agree, "lodestar" is sort of a nebulous term now. I think `lodestar-beacon` makes sense.
Related, we could move cli into a separate package and reexport all relevant objects with a 'convenient' interface in the `lodestar` package.

File: packages/lodestar/src/validator/nodejs.ts (hunk 3 above)
Comment (wemeetagain) [reply packages/lodestar/src/validator/nodejs.ts:1]: if we rename lodestar to lodestar-beacon, and add a "convenience" lodestar package that re-exports everything, this code can/should go in the validator package.
Ofc needs the logger and db broken into packages.

File: packages/lodestar/src/cli/commands/validator.ts (hunk 6 above)
Comment (wemeetagain) [reply packages/lodestar/src/cli/commands/validator.ts:2]: CLI in a separate package would be nice to separate concerns. Debugging is always a little trickier with more packages, but the monorepo makes debugging easier.
Lerna should handle the linking, just rebuild the needed packages.

**Review by ec2 — APPROVED** [review]
Other than the stuff we're discussing, which should be in another PR, LGTM.

**Review by wemeetagain — APPROVED** [review]

**Pull request merged**